packages/ethereum/contracts/ERC777/ERC777Snapshot.sol

```diff
@@ -14,171 +11,147 @@ import "@openzeppelin/contracts/token/ERC777/ERC777.sol";
 
     /**
-     * @dev Emitted by {_snapshot} when a snapshot identified by `id` is created.
+     * @dev `Snapshot` is the structure that attaches a block number to a
+     * given value, the block number attached is the one that last changed the
+     * value
      */
-    event Snapshot(uint256 id);
+    struct Snapshot {
+        // `fromBlock` is the block number that the value was generated from
+        uint128 fromBlock;
+        // `value` is the amount of tokens at a specific block number
+        uint128 value;
+    }
 
-    /**
-     * @dev Creates a new snapshot and returns its snapshot id.
-     *
-     * Emits a {Snapshot} event that contains the same id.
-     *
-     * {_snapshot} is `internal` and you have to decide how to expose it externally. Its usage may be restricted to a
-     * set of accounts, for example using {AccessControl}, or it may be open to the public.
-     *
-     * [WARNING]
-     * ====
-     * While an open way of calling {_snapshot} is required for certain trust minimization mechanisms such as forking,
-     * you must consider that it can potentially be used by attackers in two ways.
-     *
-     * First, it can be used to increase the cost of retrieval of values from snapshots, although it will grow
-     * logarithmically thus rendering this attack ineffective in the long term. Second, it can be used to target
-     * specific accounts and increase the cost of ERC777 transfers for them, in the ways specified in the Gas Costs
-     * section above.
-     *
-     * We haven't measured the actual numbers; if this is something you're interested in please reach out to us.
-     * ====
-     */
-    function _snapshot() internal virtual returns (uint256) {
-        _currentSnapshotId.increment();
+    // `accountSnapshots` is the map that tracks the balance of each address, in this
+    //  contract when the balance changes the block number that the change
+    //  occurred is also included in the map
+    mapping (address => Snapshot[]) public accountSnapshots;
 
-        uint256 currentId = _currentSnapshotId.current();
-        emit Snapshot(currentId);
-        return currentId;
-    }
+    // Tracks the history of the `totalSupply` of the token
+    Snapshot[] public totalSupplySnapshots;
 
     /**
-     * @dev Retrieves the balance of `account` at the time `snapshotId` was created.
+     * @dev Queries the balance of `_owner` at a specific `_blockNumber`
+     * @param _owner The address from which the balance will be retrieved
+     * @param _blockNumber The block number when the balance is queried
+     * @return The balance at `_blockNumber`
      */
-    function balanceOfAt(address account, uint256 snapshotId) public view returns (uint256) {
-        (bool snapshotted, uint256 value) = _valueAt(snapshotId, _accountBalanceSnapshots[account]);
-
-        return snapshotted ? value : balanceOf(account);
+    function balanceOfAt(address _owner, uint256 _blockNumber) external view returns (uint) {
```

Comment:
Shouldn't `_blockNumber` be `uint128`?

Reply:
Changing to `uint128`, since there is no standard on the snapshots, we can be more flexible here and catch early any casting issues.